include/madwf_ml.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <vector>
+#include <random>
+#include <unordered_map>
+
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <polynomial.h>
+
+#include <random_quda.h>
+#include <tune_quda.h>
+
+#include <device_vector.h>
+#include <invert_quda.h>
+#include <madwf_param.h>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+    /**
+      @brief defines the types of 5th dimension transfer matrices:
+      - Wilson: has both the color and spin indices, d.o.f = (4 x 3) * (4 x 3) = 144
+      - Spin: only has the spin index, d.o.f = (4) * (4) = 16
+      - Chiral: only has the chiral index, d.o.f = 2
+    */
+    enum class transfer_5D_t { Wilson = 144, Spin = 16, Chiral = 2 };
+
+    /**
+      @brief A helper class to instantiate the precisions (half and quarter) for madwf_tensor and madwf_transfer
+    */
+    template <template <class T> class F, class... Args>
+    void instantiate_madwf(ColorSpinorField &out, const ColorSpinorField &in, Args... args)
+    {
+      switch (checkPrecision(out, in)) {
+      case QUDA_HALF_PRECISION: {
+        if constexpr (static_cast<bool>(QUDA_PRECISION & 2)) {
+          F<short> w(out, in, args...);
+        } else {
+          errorQuda("QUDA_PRECISION=%d does not enable half precision", QUDA_PRECISION);
+        }
+      } break;
+      case QUDA_QUARTER_PRECISION: {
+        if constexpr (static_cast<bool>(QUDA_PRECISION & 1)) {
+          F<int8_t> w(out, in, args...);
+        } else {
+          errorQuda("QUDA_PRECISION=%d does not enable quarter precision", QUDA_PRECISION);
+        }
+      } break;
+      default: errorQuda("Unsupported precision %d", in.Precision());
+      }
+    }
+  } // namespace madwf_ml
+
+  /**
+    @brief A class for performing the Mobius Accelerated Domain Wall Fermion (MADWF) with the machine learning (ML)
+    version, i.e. the transfer matrix between the vectors with different Ls are _trained_.
+    - As indicated, the parameters, ultimately contained in `device_param`, need to be trained by calling the
+    `train` method.
+    - Once trained, the accelerated operator can be applied with the `apply` method.
+    @param transfer_float the floating-point type we use for the transfer matrix
+    @param trained whether this object has been trained or not
+    @param mu internal copy of the suppression factor
+    @param forward_tmp, backward_tmp persistent ColorSpinorField buffers for reuse
+    @param prec_precondition the underlying preconditioning precision
+    @param host_training_param_cache a static map that holds the cached parameters, so we do not need to load the same
+      set of parameters multiple times from disk
+  */
+  struct MadwfAcc {
+
+    using transfer_float = float;
```

Comment:
Is this abstraction necessary? More specifically, can it just be `float`, or is there a perf penalty to just having it be `double`?